cmd/junitreport/pkg/parser/stack/parser.go

```diff
@@ -0,0 +1,95 @@
+package stack
+
+import (
+	"bufio"
+
+	"github.com/openshift/origin/cmd/junitreport/pkg/api"
+	"github.com/openshift/origin/cmd/junitreport/pkg/builder"
+	"github.com/openshift/origin/cmd/junitreport/pkg/parser"
+)
+
+// NewParser returns a new parser that's capable of parsing Go unit test output
+func NewParser(builder builder.TestSuitesBuilder, testParser parser.TestDataParser, suiteParser parser.TestSuiteDataParser) parser.TestOutputParser {
+	return &testOutputParser{
+		builder:     builder,
+		testParser:  testParser,
+		suiteParser: suiteParser,
+	}
+}
+
+// testOutputParser uses a stack to parse test output. Critical assumptions that this parser makes are:
+//   1 - packages may be nested but tests may not
+//   2 - no package declarations will occur within the boundaries of a test
+//   3 - all tests and packages are fully bounded by a start and result line
+//   4 - if a package or test declaration occurs after the start of a package but before it's result,
+//       the sub-package's or member test's result line will occur before that of the parent package
+//       i.e. any test or package overlap will necessarily mean that one package's lines are a superset
+//       of any lines of tests or other packages overlapping with it
+//   5 - any text in the input file that doesn't match the above regex is necessarily the output of the
+//       current test
+type testOutputParser struct {
+	builder builder.TestSuitesBuilder
+
+	testParser  parser.TestDataParser
+	suiteParser parser.TestSuiteDataParser
+}
+
+// Parse parses output syntax of a specific class, the assumptions of which are outlined in the struct definition.
+// The specific boundary markers and metadata encoding is free to vary as long as it can be expressed as
+func (p *testOutputParser) Parse(input *bufio.Scanner) (*api.TestSuites, error) {
+	inProgress := NewTestSuiteDataStack()
+
+	currentTestData := &parser.TestData{}
+
+	for input.Scan() {
+		line := input.Text()
+
```

Comment:
with this many fallthroughs, a switch is actually harder to follow, sequential ifs would probably be easier